charmcraft/commands/build.py

```diff
@@ -182,65 +183,64 @@ def run(self, bases_indices: Optional[List[int]] = None) -> List[str]:
+                        charm_name = self.build_charm(bases_config)
                     else:
-                        matches, reason = is_base_providable(build_on)
-
-                    if matches:
-                        logger.debug(
-                            "Building for 'bases[%d]' as host matches 'build-on[%d]'.",
-                            bases_index,
-                            build_on_index,
-                        )
-                        if is_charmcraft_running_in_managed_mode():
-                            charm_name = self.build_charm(bases_config)
-                        else:
-                            with launched_environment(
-                                charm_name=self.metadata.name,
-                                project_path=self.charmdir,
-                                base=build_on,
-                                bases_index=bases_index,
-                                build_on_index=build_on_index,
-                            ) as instance:
-                                charm_name = self.pack_charm_in_instance(
-                                    instance, bases_index
-                                )
-
-                        charms.append(charm_name)
-                        break
-                    else:
-                        logger.debug(
-                            "Host does not match 'bases[%d].build-on[%d]' (%s)",
-                            bases_index,
-                            build_on_index,
-                            reason,
-                        )
+                        with launched_environment(
+                            charm_name=self.metadata.name,
+                            project_path=self.charmdir,
+                            base=build_on,
+                            bases_index=bases_index,
+                            build_on_index=build_on_index,
+                        ) as instance:
+                            charm_name = self.pack_charm_in_instance(
+                                instance, bases_index
+                            )
+
+                    charms.append(charm_name)
+                    break
                 else:
-                    logger.warning(
-                        "No suitable 'build-on' environment found in 'bases[%d]' configuration.",
+                    logger.debug(
+                        "Host does not match 'bases[%d].build-on[%d]' (%s)",
                         bases_index,
+                        build_on_index,
+                        reason,
                     )
-
-            if not charms:
-                raise CommandError(
-                    "No suitable 'build-on' environment found in any 'bases' configuration."
+            else:
+                logger.warning(
+                    "No suitable 'build-on' environment found in 'bases[%d]' configuration.",
+                    bases_index,
                 )
 
-        else:
-            charm_name = self.build_charm(None)
```

Comment:
`build_charm` and `create_manifest` shouldn't accept None as bases anymore...

Reply:
agreed

Reply:
Done!